### PR TITLE
Core - HIP buffer transfer

### PR DESCRIPTION
**Pull request by swetha097**
Cherry-picked commit - HIP Buffer Transfer

### CODE REVIEW

File: amd_openvx/openvx/api/vx_api.cpp [hunk 1]
```diff
@@ -410,12 +410,11 @@ VX_API_ENTRY vx_status VX_API_CALL vxSetContextAttribute(vx_context context, vx_
...
         case VX_CONTEXT_ATTRIBUTE_AMD_HIP_DEVICE:
             if (size == sizeof(hipDevice_t)) {
-                if (context->hip_device < 0 && agoGpuHipCreateContext(context, context->hip_device) != VX_SUCCESS) {
+                if (context->hip_device_id < 0 && agoGpuHipCreateContext(context, context->hip_device_id) != VX_SUCCESS) {
                     status = VX_FAILURE;
                 }
                 else {
-                    *(int *)ptr = context->hip_device_id;
-                    status = VX_SUCCESS;
+                    status = agoGpuHipCreateContext(context, *(int *)ptr);
```
Comment (rrawther) [inline amd_openvx/openvx/api/vx_api.cpp:417]: do we need to create again if it the device is already initialized

File: rocAL/rocAL_pybind/example/coco_pipeline.py [hunk 2]
```diff
@@ -65,7 +66,7 @@ def __init__(self, batch_size, num_threads, device_id, seed, data_dir, ann_dir,
                                                 output_layout=types.NCHW,
                                                 crop=(crop, crop),
                                                 image_type=types.RGB,
-                                                mean=[0.485 * 255, 0.456 * 255, 0.406 * 255],
+                                                mean=[0.485 * 255, 0.456 *255, 0.406 * 255],
```
Comment (rrawther) [inline rocAL/rocAL_pybind/example/coco_pipeline.py:69]: please revert it to original code

File: rocAL/rocAL_pybind/example/coco_pipeline.py [hunk 3]
```diff
@@ -145,10 +149,36 @@ def __init__(self, pipelines, tensor_layout=types.NCHW, reverse_channels=False,
         # Image sizes of a batch
         self.img_size = np.zeros((self.bs * 2), dtype="int32")
 
+    def get_cupy_array_from_address(self):
+        if(self.tensor_format == types.NCHW):
+            address = self.loader.copyToHipTensorNCHW(
+                self.multiplier, self.offset, self.reverse_channels, int(self.tensor_dtype))
+            print("Address : ",address)
```
Comment (rrawther) [inline rocAL/rocAL_pybind/example/coco_pipeline.py:156]: comment out this print command if it is needed only for debug

File: rocAL/rocAL_pybind/example/coco_pipeline.py [hunk 4]
```diff
@@ -145,10 +149,36 @@ def __init__(self, pipelines, tensor_layout=types.NCHW, reverse_channels=False,
...
+            mem = cp.cuda.memory.UnownedMemory(address, 0, None, self.device_id)
```
Comment (rrawther) [inline rocAL/rocAL_pybind/example/coco_pipeline.py:157]: Why the size is 0 here

File: rocAL/rocAL_pybind/example/coco_pipeline.py [hunk 5]
```diff
@@ -145,10 +149,36 @@ def __init__(self, pipelines, tensor_layout=types.NCHW, reverse_channels=False,
...
+                    (self.bs*self.n, int(self.h/self.bs), self.w,  self.p), "float32", memptr)
+            else:
+                self.hip_array = cp.ndarray(
+                    (self.bs*self.n, int(self.h/self.bs), self.w,  self.p), "float16", memptr)
+
     def next(self):
         return self.__next__()
 
     def __next__(self):
+        print("In the next routine of COCO Iterator")
```
Comment (rrawther) [inline rocAL/rocAL_pybind/example/coco_pipeline.py:181]: comment print statement

File: rocAL/rocAL_pybind/example/coco_pipeline.py [hunk 6]
```diff
@@ -162,12 +192,15 @@ def __next__(self):
...
+            if(types.NCHW == self.tensor_format):
+                self.loader.copyToTensorNCHW(
+                    self.out, self.multiplier, self.offset, self.reverse_channels, int(self.tensor_dtype))
+            else:
+                self.loader.copyToTensorNHWC(
+                    self.out, self.multiplier, self.offset, self.reverse_channels, int(self.tensor_dtype))
         else:
-            self.loader.copyToTensorNHWC(
-                self.out, self.multiplier, self.offset, self.reverse_channels, int(self.tensor_dtype))
+            self.get_cupy_array_from_address()
```
Comment (rrawther) [inline rocAL/rocAL_pybind/example/coco_pipeline.py:203]: It is better to call copy fun here and pass the pointer to get_cupy_array..()

File: rocAL/rocAL_pybind/example/coco_pipeline.py [hunk 7]
```diff
@@ -197,24 +230,36 @@ def __next__(self):
...
-        elif self.tensor_dtype == types.FLOAT16:
-            return torch.from_numpy(self.out.astype(np.float16)), encoded_bboxes_tensor, encodded_labels_tensor, image_id_tensor, image_size_tensor
+        if self.device == "cpu":
+            if self.tensor_dtype == types.FLOAT:
+                return torch.from_numpy(self.out), encoded_bboxes_tensor, encodded_labels_tensor, image_id_tensor, image_size_tensor
+            elif self.tensor_dtype == types.FLOAT16:
+                return torch.from_numpy(self.out.astype(np.float16)), encoded_bboxes_tensor, encodded_labels_tensor, image_id_tensor, image_size_tensor
+
+        if self.device == "cuda":
+            dlpack_array = self.hip_array.toDlpack()
```
Comment (rrawther) [inline rocAL/rocAL_pybind/example/coco_pipeline.py:247]: Eventually we need to get rid of all these calls and do the conversion to numpy in the pybind wrapper function.

File: rocAL/rocAL_pybind/rali_pybind.cpp [hunk 8]
```diff
@@ -83,6 +83,28 @@ namespace rali{
         return py::cast<py::none>(Py_None);
     }
 
+    py::object wrapper_hip_tensor32(RaliContext context, RaliTensorLayout tensor_format,
```
Comment (rrawther) [inline rocAL/rocAL_pybind/rali_pybind.cpp:86]: rename to reflect the copy routine

File: rocAL/rocAL_pybind/setup.py [hunk 9]
```diff
@@ -28,7 +28,7 @@ def run(self):
     def build_extension(self, ext):
         extdir = os.path.abspath(os.path.dirname(self.get_ext_fullpath(ext.name)))
         cmake_args = ['-DCMAKE_LIBRARY_OUTPUT_DIRECTORY=' + extdir,
-                      '-DPYTHON_EXECUTABLE=' + sys.executable, '-DGPU_SUPPORT=ON', '-DBACKEND=OPENCL']
+                      '-DPYTHON_EXECUTABLE=' + sys.executable, '-DGPU_SUPPORT=ON', '-DBACKEND=HIP'] #Needs to be changed back to OCL for OCL backend
```
Comment (rrawther) [inline rocAL/rocAL_pybind/setup.py:31]: add option in CMake to take it from command line

File: tests/vision_tests/runVisionTests.py [hunk 10]
```diff
@@ -516,9 +516,9 @@ def multiCaseProfilerOCL(nodeList, case_num_list):
         os.system('echo '+echo1 +
```
Comment (rrawther) [inline tests/vision_tests/runVisionTests.py:516]: this shouldn't be part of this PR?

File: tests/vision_tests/rocprof_counters.txt [hunk 11]
```diff
@@ -0,0 +1,4 @@
+pmc: Wavefronts VALUInsts VFetchInsts VWriteInsts VALUUtilization VALUBusy WriteSize
```
Comment (rrawther) [inline tests/vision_tests/rocprof_counters.txt:1]: this is not part of this PR

**Review by rrawther — CHANGES_REQUESTED** [review]
Please address review comments

File: tests/vision_tests/rocprof_counters.txt (hunk 11 above)
Comment (swetha097) [reply tests/vision_tests/rocprof_counters.txt]: Reverted this commit

File: tests/vision_tests/runVisionTests.py (hunk 10 above)
Comment (swetha097) [reply tests/vision_tests/runVisionTests.py]: Reverted this commit

File: amd_openvx/openvx/api/vx_api.cpp (hunk 1 above)
Comment (asalmanp) [inline amd_openvx/openvx/api/vx_api.cpp:417]: What's the reason for this change? if the device is already initialized above then you don't need to call agoGpuHipCreateContext again here

File: rocAL/rocAL_pybind/setup.py (hunk 9 above)
Comment (swetha097) [reply rocAL/rocAL_pybind/setup.py]: We are now able to pass the user given cmd line args in run.sh & use the same in setup.py file to build either the OCL / HIP backends.
The HIP is added as the default backend while running the run.sh file

File: rocAL/rocAL_pybind/example/coco_pipeline.py (hunk 4 above)
Comment (swetha097) [reply rocAL/rocAL_pybind/example/coco_pipeline.py:157]: We are wrap our pointer using CuPy's UnownedMemory to get the memory pointer of type cupy.cuda.MemoryPointer required to create the cupy ndarray
The memptr points to the array content head 
The "mem" is not used anywhere except to get the pointer, hence size was given as "0"

File: rocAL/rocAL_pybind/example/coco_pipeline.py [hunk 12]
```diff
@@ -11,14 +11,16 @@
...
+import torch.utils.dlpack as dp
 
 
 class COCOPipeline(Pipeline):
     def __init__(self, batch_size, num_threads, device_id, seed, data_dir, ann_dir, default_boxes,  crop, rali_cpu=True, display=False):
         super(COCOPipeline, self).__init__(batch_size, num_threads,
                                            device_id, seed=seed, rali_cpu=rali_cpu)
         self.input = ops.COCOReader(
-            file_root=data_dir, annotations_file=ann_dir, random_shuffle=True, seed=seed)
+            file_root=data_dir, annotations_file=ann_dir,random_shuffle=True,seed=seed)
```
[the remark this answers is not in the document]
Comment (swetha097) [reply rocAL/rocAL_pybind/example/coco_pipeline.py]: Corrected the formatting

File: rocAL/rocAL_pybind/example/coco_pipeline.py (hunk 2 above)
Comment (swetha097) [reply rocAL/rocAL_pybind/example/coco_pipeline.py]: Corrected the formatting

File: rocAL/rocAL_pybind/example/coco_pipeline.py (hunk 3 above)
Comment (swetha097) [reply rocAL/rocAL_pybind/example/coco_pipeline.py]: Done

File: rocAL/rocAL_pybind/example/coco_pipeline.py (hunk 6 above)
Comment (swetha097) [reply rocAL/rocAL_pybind/example/coco_pipeline.py]: Made this change

File: amd_openvx/openvx/api/vx_api.cpp (hunk 1 above)
Comment (swetha097) [reply amd_openvx/openvx/api/vx_api.cpp]: Corrected it.
We create the hip context just once now.

Comment (swetha097) [reply amd_openvx/openvx/api/vx_api.cpp]: Corrected it.
We create the hip context just once now.

File: rocAL/rocAL/rocAL_hip/rali_hip_kernels.cpp [hunk 13]
```diff
@@ -228,7 +228,7 @@ int HipExecCopyInt8ToNHWC
 int HipExecCopyInt8ToNCHW
 (
     hipStream_t stream,
-    void*     inp_image_u8,
+    const void*     inp_image_u8,
```
[the remark this answers is not in the document]
Comment (swetha097) [reply rocAL/rocAL/rocAL_hip/rali_hip_kernels.cpp:231]: const is used for input image in other variants

File: rocAL/rocAL_pybind/rali_pybind.cpp (hunk 8 above)
Comment (swetha097) [reply rocAL/rocAL_pybind/rali_pybind.cpp:86]: Done

File: rocAL/rocAL/include/master_graph.h [hunk 14]
```diff
@@ -58,6 +58,8 @@ class MasterGraph
     Status copy_output(void* out_ptr, size_t out_size);
     Status copy_out_tensor_planar(void *out_ptr, RaliTensorFormat format, float multiplier0, float multiplier1, float multiplier2,
                     float offset0, float offset1, float offset2, bool reverse_channels, RaliTensorDataType output_data_type);
+    void *copy_hip_out_tensor(RaliTensorFormat format, float multiplier0, float multiplier1, float multiplier2,
```
[the remark this answers is not in the document]
Comment (swetha097) [reply rocAL/rocAL/include/master_graph.h:61]: Done

**Comment by swetha097** [comment]
> Please address review comments

I have addressed the comments @rrawther

File: amd_openvx/openvx/ago/ago_util_hip.cpp [hunk 15]
```diff
@@ -27,6 +27,9 @@ THE SOFTWARE.
 // Create HIP Context
 int agoGpuHipCreateContext(AgoContext *context, int deviceID) {
     if (deviceID >= 0) {
```
Comment (rrawther) [inline amd_openvx/openvx/ago/ago_util_hip.cpp:29]: This is already in top? remove it from this PR

File: amd_openvx/openvx/api/vx_api.cpp [hunk 16]
```diff
@@ -410,12 +410,11 @@ VX_API_ENTRY vx_status VX_API_CALL vxSetContextAttribute(vx_context context, vx_
 #elif ENABLE_HIP
         case VX_CONTEXT_ATTRIBUTE_AMD_HIP_DEVICE:
             if (size == sizeof(hipDevice_t)) {
-                if (context->hip_device < 0 && agoGpuHipCreateContext(context, context->hip_device) != VX_SUCCESS) {
-                    status = VX_FAILURE;
+                if (*(int *)ptr > 0 && agoGpuHipCreateContext(context, *(int *)ptr) == VX_SUCCESS) {
```
Comment (rrawther) [inline amd_openvx/openvx/api/vx_api.cpp:413]: please use local variable device_id = *(int *)ptr and use that within the if statement

File: rocAL/rocAL/include/video_loader.h [hunk 17]
```diff
@@ -57,6 +57,8 @@ class VideoLoader : public VideoLoaderModule
     void set_prefetch_queue_depth(size_t prefetch_queue_depth) override;
     std::vector<size_t> get_sequence_start_frame_number();
     std::vector<std::vector<float>> get_sequence_frame_timestamps();
+    void shut_down() override;
```
Comment (rrawther) [inline rocAL/rocAL/include/video_loader.h:60]: Looks like, you haven't merged the tot. Many changes which is already there in tot is showing here

**Review by rrawther — COMMENTED** [review]
@swetha097 please make changes to develop. The PR is showing many merged changes.

File: CHANGELOG.md [hunk 18]
```diff
@@ -11,6 +11,7 @@
 ### Added
```
Comment (rrawther) [inline CHANGELOG.md:11]: already there in tot

File: CHANGELOG.md [hunk 19]
```diff
@@ -11,6 +11,7 @@
...
 
 * Support for cmake 3.22.X
+* Support for mv_compile with the HIP GPU backend
```
Comment (rrawther) [inline CHANGELOG.md:14]: already there in tot

File: rocAL/rocAL/source/master_graph.cpp [hunk 20]
```diff
@@ -1511,3 +1511,53 @@ MasterGraph::copy_out_tensor_planar(void *out_ptr, RaliTensorFormat format, floa
     _convert_time.end();
     return Status::OK;
 }
+
+void *
+MasterGraph::copy_hip_out_tensor(RaliTensorFormat format, float multiplier0, float multiplier1,
```
Comment (rrawther) [inline rocAL/rocAL/source/master_graph.cpp:1516]: todo:: we need to have only one function which should work for both host and device

File: rocAL/rocAL_pybind/example/coco_pipeline.py [hunk 21]
```diff
@@ -145,6 +150,16 @@ def __init__(self, pipelines, tensor_layout=types.NCHW, reverse_channels=False,
         # Image sizes of a batch
         self.img_size = np.zeros((self.bs * 2), dtype="int32")
 
+    def get_cupy_array_from_address(self, address):
```
Comment (rrawther) [inline rocAL/rocAL_pybind/example/coco_pipeline.py:153]: Can you update with the new way of not using cupy

File: rocAL/rocAL_pybind/run.sh [hunk 22]
```diff
@@ -1,17 +1,42 @@
 #!/bin/bash
 
-DEFAULT_PYTHON=$(which python3.6) ## Gets the default python
+PYTHON_VERSION=`python3 -c "import sys;t='{v[0]}.{v[1]}'.format(v=list(sys.version_info[:2]));sys.stdout.write(t)";` ##Gets the Python3 version
```
Comment (rrawther) [inline rocAL/rocAL_pybind/run.sh:3]: please fix all codacy errors

File: utilities/CMakeLists.txt [hunk 23]
```diff
@@ -25,9 +25,9 @@ list(APPEND CMAKE_MODULE_PATH ${PROJECT_SOURCE_DIR}/../amd_openvx/cmake)
 #set(ROCM_PATH /opt/rocm CACHE PATH " default ROCm installation path")
 
 if(GPU_SUPPORT)
+    find_package(FFmpeg QUIET)
```
Comment (rrawther) [inline utilities/CMakeLists.txt:28]: I think this file is not part of this PR

**Review by rrawther — COMMENTED** [review]
@swetha097 : I still see lot of files from tot in this PR. May be you can create a new PR which doesn't use cupy with all the changes requested in this PR

**Comment by kiritigowda** [comment]
@swetha097 this PR has merge conflicts, please sync with TOT master

**Comment by swetha097** [comment]
> a new PR which doesn't use cupy with

Sure @rrawther 
I will create a new PR which has eliminated the cupy usage & delete this PR

**Comment by swetha097** [comment]
> @swetha097 this PR has merge conflicts, please sync with TOT master

I will create a new PR & delete this

**Comment by swetha097** [comment]
Closing this PR since there are files from TOT.
Will create a new PR for the same
